File: sdks/go/cmd/elydora/plugins/cline_contract.go
package plugins

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

const (
	clineAgentKey        = "cline"
	clineGuardFileName   = "PreToolUse.mjs"
	clineAuditFileName   = "PostToolUse.mjs"
	clineGuardScript     = "guard.js"
	clineAuditScript     = "hook.js"
	clineMetadataMarker  = "// @elydora-cline-hook "
	clineMetadataVersion = 1
)

type clineHookMetadata struct {
	Version     int    `json:"version"`
	Kind        string `json:"kind"`
	AgentID     string `json:"agentId"`
	RuntimePath string `json:"runtimePath"`
}

type clineHookFile struct {
	exists   bool
	filePath string
	source   string
	metadata *clineHookMetadata
}

type clineHookPaths struct {
	hooksDirectory string
	guardPath      string
	auditPath      string
}

type clineRuntimeContract struct {
	agentID        string
	agentDirectory string
	guardPath      string
	auditPath      string
}

func clineHomeDirectory() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("resolve home directory: %w", err)
	}
	return home, nil
}

func clineElydoraRoot() (string, error) {
	home, err := clineHomeDirectory()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, ".elydora"), nil
}

func resolveClineHookFiles() (clineHookPaths, error) {
	clineDirectory := strings.TrimSpace(os.Getenv("CLINE_DIR"))
	if clineDirectory == "" {
		home, err := clineHomeDirectory()
		if err != nil {
			return clineHookPaths{}, err
		}
		clineDirectory = filepath.Join(home, ".cline")
	} else {
		absolute, err := filepath.Abs(clineDirectory)
		if err != nil {
			return clineHookPaths{}, fmt.Errorf("resolve CLINE_DIR path: %w", err)
		}
		clineDirectory = absolute
	}
	hooksDirectory := filepath.Join(clineDirectory, "hooks")
	return clineHookPaths{
		hooksDirectory: hooksDirectory,
		guardPath:      filepath.Join(hooksDirectory, clineGuardFileName),
		auditPath:      filepath.Join(hooksDirectory, clineAuditFileName),
	}, nil
}

func sameClineAgentID(left, right string) bool {
	if runtime.GOOS == "windows" {
		return strings.EqualFold(left, right)
	}
	return left == right
}

func normalizeClinePath(value string) (string, error) {
	absolute, err := filepath.Abs(value)
	if err != nil {
		return "", err
	}
	normalized := filepath.Clean(absolute)
	if runtime.GOOS == "windows" {
		normalized = strings.ToLower(normalized)
	}
	return normalized, nil
}

func sameClinePath(left, right string) (bool, error) {
	normalizedLeft, err := normalizeClinePath(left)
	if err != nil {
		return false, err
	}
	normalizedRight, err := normalizeClinePath(right)
	if err != nil {
		return false, err
	}
	return normalizedLeft == normalizedRight, nil
}

func buildClineMetadata(kind, agentID, runtimePath string) (clineHookMetadata, error) {
	metadata := clineHookMetadata{
		Version:     clineMetadataVersion,
		Kind:        kind,
		AgentID:     agentID,
		RuntimePath: runtimePath,
	}
	if err := validateClineMetadata(metadata); err != nil {
		return clineHookMetadata{}, err
	}
	return metadata, nil
}

func validateClineMetadata(metadata clineHookMetadata) error {
	if metadata.Version != clineMetadataVersion {
		return fmt.Errorf("metadata version must be 1")
	}
	if metadata.Kind != "guard" && metadata.Kind != "audit" {
		return fmt.Errorf(`metadata kind must be "guard" or "audit"`)
	}
	if metadata.AgentID == "" {
		return fmt.Errorf("metadata agentId must be a non-empty string")
	}
	if !filepath.IsAbs(metadata.RuntimePath) {
		return fmt.Errorf("metadata runtimePath must be an absolute path")
	}
	return nil
}

func encodeClineMetadata(metadata clineHookMetadata) (string, error) {
	encoded, err := json.Marshal(metadata)
	if err != nil {
		return "", fmt.Errorf("encode Cline hook metadata: %w", err)
	}
	return base64.RawURLEncoding.EncodeToString(encoded), nil
}

func clineJSONString(value string) (string, error) {
	encoded, err := json.Marshal(value)
	if err != nil {
		return "", err
	}
	return string(encoded), nil
}

func buildClineWrapper(metadata clineHookMetadata) (string, error) {
	if err := validateClineMetadata(metadata); err != nil {
		return "", err
	}
	encodedMetadata, err := encodeClineMetadata(metadata)
	if err != nil {
		return "", err
	}
	hookKind, err := clineJSONString(metadata.Kind)
	if err != nil {
		return "", fmt.Errorf("encode Cline hook kind: %w", err)
	}
	runtimePath, err := clineJSONString(metadata.RuntimePath)
	if err != nil {
		return "", fmt.Errorf("encode Cline runtime path: %w", err)
	}
	const template = `#!/usr/bin/env node
__METADATA_LINE__
import { spawn } from 'node:child_process';

const hookKind = __HOOK_KIND__;
const runtimePath = __RUNTIME_PATH__;

function runRuntime(input) {
  return new Promise((resolve, reject) => {
    const child = spawn(process.execPath, [runtimePath], {
      stdio: ['pipe', 'pipe', 'pipe'],
      windowsHide: true,
    });
    const stdout = [];
    const stderr = [];
    let stdinError;
    child.stdout.on('data', (chunk) => stdout.push(chunk));
    child.stderr.on('data', (chunk) => stderr.push(chunk));
    child.stdin.on('error', (error) => {
      if (error?.code !== 'EPIPE') stdinError = error;
    });
    child.once('error', reject);
    child.once('close', (code, signal) => {
      if (stdinError) {
        reject(stdinError);
        return;
      }
      resolve({
        code,
        signal,
        stdout: Buffer.concat(stdout).toString('utf-8'),
        stderr: Buffer.concat(stderr).toString('utf-8'),
      });
    });
    child.stdin.end(input);
  });
}

async function main() {
  const chunks = [];
  for await (const chunk of process.stdin) chunks.push(chunk);
  const result = await runRuntime(Buffer.concat(chunks));
  if (result.stdout) process.stderr.write(result.stdout);
  if (result.stderr) process.stderr.write(result.stderr);

  if (hookKind === 'guard' && result.code === 2) {
    const errorMessage = result.stderr.trim() || 'Agent is frozen by Elydora.';
    process.stdout.write('HOOK_CONTROL\t' + JSON.stringify({ cancel: true, errorMessage }) + '\n');
    return;
  }
  if (result.signal) throw new Error('runtime terminated by signal ' + result.signal);
  if (result.code !== 0) throw new Error('runtime exited with code ' + result.code);
}

main().catch((error) => {
  const message = error instanceof Error ? error.stack || error.message : String(error);
  process.stderr.write('Elydora Cline ' + hookKind + ' wrapper failed: ' + message + '\n');
  process.exitCode = 1;
});
`
	return strings.NewReplacer(
		"__METADATA_LINE__", clineMetadataMarker+encodedMetadata,
		"__HOOK_KIND__", hookKind,
		"__RUNTIME_PATH__", runtimePath,
	).Replace(template), nil
}

func decodeClineMetadata(encoded string) (clineHookMetadata, error) {
	raw, err := base64.RawURLEncoding.DecodeString(encoded)
	if err != nil {
		return clineHookMetadata{}, fmt.Errorf("invalid encoded metadata: %w", err)
	}
	var fields map[string]json.RawMessage
	if err := json.Unmarshal(raw, &fields); err != nil {
		return clineHookMetadata{}, fmt.Errorf("invalid encoded metadata: %w", err)
	}
	expected := []string{"version", "kind", "agentId", "runtimePath"}
	if len(fields) != len(expected) {
		return clineHookMetadata{}, fmt.Errorf("metadata contains an unexpected field set")
	}
	for _, field := range expected {
		if _, exists := fields[field]; !exists {
			return clineHookMetadata{}, fmt.Errorf("metadata contains an unexpected field set")
		}
	}
	var metadata clineHookMetadata
	if err := json.Unmarshal(raw, &metadata); err != nil {
		return clineHookMetadata{}, fmt.Errorf("invalid encoded metadata: %w", err)
	}
	if err := validateClineMetadata(metadata); err != nil {
		return clineHookMetadata{}, err
	}
	return metadata, nil
}

func parseClineMetadata(filePath, source string) (*clineHookMetadata, error) {
	lines := strings.SplitN(source, "\n", 3)
	if len(lines) < 2 {
		return nil, nil
	}
	markerLine := strings.TrimSuffix(lines[1], "\r")
	if !strings.HasPrefix(markerLine, clineMetadataMarker) {
		return nil, nil
	}
	metadata, err := decodeClineMetadata(strings.TrimPrefix(markerLine, clineMetadataMarker))
	if err != nil {
		return nil, fmt.Errorf("parse Elydora Cline hook metadata at %s: %w", filePath, err)
	}
	return &metadata, nil
}

func assertClineWrapperIntegrity(file clineHookFile) error {
	if file.metadata == nil {
		return nil
	}
	expected, err := buildClineWrapper(*file.metadata)
	if err != nil {
		return err
	}
	if file.source != expected {
		return fmt.Errorf("Elydora Cline hook at %s does not match the managed template", file.filePath)
	}
	return nil
}

func validateClineAgentSegment(agentID string) error {
	if agentID == "." || agentID == ".." || filepath.Base(agentID) != agentID {
		return fmt.Errorf("Elydora Cline hook metadata contains an invalid agentId")
	}
	return nil
}

func clineContractForFiles(
	guardFile, auditFile clineHookFile,
) (*clineRuntimeContract, error) {
	if guardFile.metadata == nil || auditFile.metadata == nil {
		return nil, nil
	}
	if err := assertClineWrapperIntegrity(guardFile); err != nil {
		return nil, err
	}
	if err := assertClineWrapperIntegrity(auditFile); err != nil {
		return nil, err
	}
	guard := guardFile.metadata
	audit := auditFile.metadata
	if guard.Kind != "guard" || audit.Kind != "audit" {
		return nil, fmt.Errorf("Elydora Cline hook files contain mismatched event metadata")
	}
	if !sameClineAgentID(guard.AgentID, audit.AgentID) {
		return nil, fmt.Errorf("Elydora Cline hook files reference different agents")
	}
	if err := validateClineAgentSegment(guard.AgentID); err != nil {
		return nil, err
	}
	runtimeRoot, err := clineElydoraRoot()
	if err != nil {
		return nil, err
	}
	agentDirectory := filepath.Join(runtimeRoot, guard.AgentID)
	expectedGuard := filepath.Join(agentDirectory, clineGuardScript)
	expectedAudit := filepath.Join(agentDirectory, clineAuditScript)
	guardMatches, err := sameClinePath(guard.RuntimePath, expectedGuard)
	if err != nil {
		return nil, fmt.Errorf("compare Cline guard runtime path: %w", err)
	}
	auditMatches, err := sameClinePath(audit.RuntimePath, expectedAudit)
	if err != nil {
		return nil, fmt.Errorf("compare Cline audit runtime path: %w", err)
	}
	if !guardMatches || !auditMatches {
		return nil, fmt.Errorf("Elydora Cline hook metadata references an unexpected runtime path")
	}
	return &clineRuntimeContract{
		agentID:        guard.AgentID,
		agentDirectory: agentDirectory,
		guardPath:      expectedGuard,
		auditPath:      expectedAudit,
	}, nil
}

File: sdks/node/src/plugins/cline-contract.ts
import os from 'node:os';
import path from 'node:path';

export const AGENT_KEY = 'cline';
export const GUARD_SCRIPT = 'guard.js';
export const AUDIT_SCRIPT = 'hook.js';
export const GUARD_FILE_NAME = 'PreToolUse.mjs';
export const AUDIT_FILE_NAME = 'PostToolUse.mjs';

const METADATA_MARKER = '// @elydora-cline-hook ';
const METADATA_VERSION = 1;

export type ClineHookKind = 'guard' | 'audit';

export interface ClineHookMetadata {
  readonly version: 1;
  readonly kind: ClineHookKind;
  readonly agentId: string;
  readonly runtimePath: string;
}

export interface ClineHookFile {
  readonly exists: boolean;
  readonly filePath: string;
  readonly source?: string;
  readonly metadata?: ClineHookMetadata;
}

export interface ClineRuntimeContract {
  readonly agentId: string;
  readonly agentDirectory: string;
  readonly guardPath: string;
  readonly auditPath: string;
}

type JsonObject = Record<string, unknown>;

function isObject(value: unknown): value is JsonObject {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

function asError(error: unknown): Error {
  return error instanceof Error ? error : new Error(String(error));
}

function samePath(left: string, right: string): boolean {
  const normalizedLeft = path.resolve(left);
  const normalizedRight = path.resolve(right);
  return process.platform === 'win32'
    ? normalizedLeft.toLowerCase() === normalizedRight.toLowerCase()
    : normalizedLeft === normalizedRight;
}

export function sameAgentId(left: string, right: string): boolean {
  return process.platform === 'win32' ? left.toLowerCase() === right.toLowerCase() : left === right;
}

export function resolveHooksDirectory(): string {
  const configuredDirectory = process.env.CLINE_DIR?.trim();
  const clineDirectory = configuredDirectory
    ? path.resolve(configuredDirectory)
    : path.join(os.homedir(), '.cline');
  return path.join(clineDirectory, 'hooks');
}

export function resolveHookFiles(): {
  readonly hooksDirectory: string;
  readonly guardPath: string;
  readonly auditPath: string;
} {
  const hooksDirectory = resolveHooksDirectory();
  return {
    hooksDirectory,
    guardPath: path.join(hooksDirectory, GUARD_FILE_NAME),
    auditPath: path.join(hooksDirectory, AUDIT_FILE_NAME),
  };
}

export function buildMetadata(
  kind: ClineHookKind,
  agentId: string,
  runtimePath: string,
): ClineHookMetadata {
  if (!agentId) throw new Error('agentId is required');
  if (!path.isAbsolute(runtimePath)) throw new Error(`Cline ${kind} runtime path must be absolute`);
  return { version: METADATA_VERSION, kind, agentId, runtimePath };
}

function encodeMetadata(metadata: ClineHookMetadata): string {
  return Buffer.from(JSON.stringify(metadata), 'utf-8').toString('base64url');
}

export function buildWrapper(metadata: ClineHookMetadata): string {
  return `#!/usr/bin/env node
${METADATA_MARKER}${encodeMetadata(metadata)}
import { spawn } from 'node:child_process';

const hookKind = ${JSON.stringify(metadata.kind)};
const runtimePath = ${JSON.stringify(metadata.runtimePath)};

function runRuntime(input) {
  return new Promise((resolve, reject) => {
    const child = spawn(process.execPath, [runtimePath], {
      stdio: ['pipe', 'pipe', 'pipe'],
      windowsHide: true,
    });
    const stdout = [];
    const stderr = [];
    let stdinError;
    child.stdout.on('data', (chunk) => stdout.push(chunk));
    child.stderr.on('data', (chunk) => stderr.push(chunk));
    child.stdin.on('error', (error) => {
      if (error?.code !== 'EPIPE') stdinError = error;
    });
    child.once('error', reject);
    child.once('close', (code, signal) => {
      if (stdinError) {
        reject(stdinError);
        return;
      }
      resolve({
        code,
        signal,
        stdout: Buffer.concat(stdout).toString('utf-8'),
        stderr: Buffer.concat(stderr).toString('utf-8'),
      });
    });
    child.stdin.end(input);
  });
}

async function main() {
  const chunks = [];
  for await (const chunk of process.stdin) chunks.push(chunk);
  const result = await runRuntime(Buffer.concat(chunks));
  if (result.stdout) process.stderr.write(result.stdout);
  if (result.stderr) process.stderr.write(result.stderr);

  if (hookKind === 'guard' && result.code === 2) {
    const errorMessage = result.stderr.trim() || 'Agent is frozen by Elydora.';
    process.stdout.write(JSON.stringify({ cancel: true, errorMessage }) + '\\n');
    return;
  }
  if (result.signal) throw new Error('runtime terminated by signal ' + result.signal);
  if (result.code !== 0) throw new Error('runtime exited with code ' + result.code);
}

main().catch((error) => {
  const message = error instanceof Error ? error.stack || error.message : String(error);
  process.stderr.write('Elydora Cline ' + hookKind + ' wrapper failed: ' + message + '\\n');
  process.exitCode = 1;
});
`;
}

function validateMetadata(value: unknown): ClineHookMetadata {
  if (!isObject(value)) throw new Error('metadata must be an object');
  const keys = Object.keys(value).sort();
  if (keys.join(',') !== 'agentId,kind,runtimePath,version') {
    throw new Error('metadata contains an unexpected field set');
  }
  if (value.version !== METADATA_VERSION) throw new Error('metadata version must be 1');
  if (value.kind !== 'guard' && value.kind !== 'audit') {
    throw new Error('metadata kind must be "guard" or "audit"');
  }
  if (typeof value.agentId !== 'string' || value.agentId.length === 0) {
    throw new Error('metadata agentId must be a non-empty string');
  }
  if (typeof value.runtimePath !== 'string' || !path.isAbsolute(value.runtimePath)) {
    throw new Error('metadata runtimePath must be an absolute path');
  }
  return value as unknown as ClineHookMetadata;
}

export function parseMetadata(filePath: string, source: string): ClineHookMetadata | undefined {
  const markerLine = source.split(/\r?\n/, 3)[1];
  if (!markerLine?.startsWith(METADATA_MARKER)) return undefined;
  try {
    const encoded = markerLine.slice(METADATA_MARKER.length);
    const decoded = Buffer.from(encoded, 'base64url').toString('utf-8');
    return validateMetadata(JSON.parse(decoded));
  } catch (error) {
    throw new Error(`Failed to parse Elydora Cline hook metadata at ${filePath}: ${errorMessage(error)}`, {
      cause: asError(error),
    });
  }
}

export function assertWrapperIntegrity(file: ClineHookFile): void {
  if (!file.metadata || file.source === undefined) return;
  if (file.source !== buildWrapper(file.metadata)) {
    throw new Error(`Elydora Cline hook at ${file.filePath} does not match the managed template`);
  }
}

function validateAgentSegment(agentId: string): void {
  if (agentId === '.' || agentId === '..' || path.basename(agentId) !== agentId) {
    throw new Error('Elydora Cline hook metadata contains an invalid agentId');
  }
}

export function runtimeContract(
  guardFile: ClineHookFile,
  auditFile: ClineHookFile,
): ClineRuntimeContract | undefined {
  if (!guardFile.metadata || !auditFile.metadata) return undefined;
  assertWrapperIntegrity(guardFile);
  assertWrapperIntegrity(auditFile);
  const guard = guardFile.metadata;
  const audit = auditFile.metadata;
  if (guard.kind !== 'guard' || audit.kind !== 'audit') {
    throw new Error('Elydora Cline hook files contain mismatched event metadata');
  }
  if (!sameAgentId(guard.agentId, audit.agentId)) {
    throw new Error('Elydora Cline hook files reference different agents');
  }
  validateAgentSegment(guard.agentId);
  const agentDirectory = path.join(os.homedir(), '.elydora', guard.agentId);
  const expectedGuard = path.join(agentDirectory, GUARD_SCRIPT);
  const expectedAudit = path.join(agentDirectory, AUDIT_SCRIPT);
  if (!samePath(guard.runtimePath, expectedGuard) || !samePath(audit.runtimePath, expectedAudit)) {
    throw new Error('Elydora Cline hook metadata references an unexpected runtime path');
  }
  return {
    agentId: guard.agentId,
    agentDirectory,
    guardPath: expectedGuard,
    auditPath: expectedAudit,
  };
}
